### PR TITLE
Introduce PoseGraphInterface.

### PR DESCRIPTION
**Pull request by cschuet**
[RFC=0005](https://github.com/googlecartographer/rfcs/blob/master/text/0005-pose-graph-interface.md)

### CODE REVIEW

File: cartographer/mapping/pose_graph_interface.h [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+  struct SubmapData {
+    std::shared_ptr<const Submap> submap;
+    transform::Rigid3d pose;
+  };
+
+  virtual ~PoseGraphInterface() {}
+
+  // Computes optimized poses.
+  virtual void RunFinalOptimization() = 0;
+
```
Comment (gaschler) [inline cartographer/mapping/pose_graph_interface.h]: Delete copy/assignment constructors (as in MapBuilderInterface)?

File: cartographer/mapping/pose_graph.h [hunk 2]
```diff
@@ -100,9 +74,6 @@ class PoseGraph {
   // Finishes the given trajectory.
   virtual void FinishTrajectory(int trajectory_id) = 0;
 
-  // Checks if the given trajectory is finished.
```
Comment (gaschler) [inline cartographer/mapping/pose_graph.h]: in line 56, style: override destructor

File: cartographer/mapping/pose_graph_interface.h [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CARTOGRAPHER_MAPPING_POSE_GRAPH_INTERFACE_H_
+#define CARTOGRAPHER_MAPPING_POSE_GRAPH_INTERFACE_H_
+
+#include "cartographer/mapping/id.h"
```
Comment (gaschler) [inline cartographer/mapping/pose_graph_interface.h]: include vector?

File: cartographer/mapping/pose_graph.h (hunk 2 above)
Comment (cschuet) [reply cartographer/mapping/pose_graph.h]: Done.

File: cartographer/mapping/pose_graph_interface.h (hunk 3 above)
Comment (cschuet) [reply cartographer/mapping/pose_graph_interface.h]: Done.

File: cartographer/mapping/pose_graph_interface.h (hunk 1 above)
Comment (cschuet) [reply cartographer/mapping/pose_graph_interface.h]: Done.

**Comment by cschuet** [comment]
@wally-the-cartographer merge

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user cschuet. Merge queue now has a length of 1.

**Comment by wally-the-cartographer** [comment]
Refusing to merge, since the combined checks state is failure

**Comment by cschuet** [comment]
@wally-the-cartographer merge

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user cschuet. Merge queue now has a length of 1.

**Pull request merged**